Analysis: Add detail response support for _analyze API

johtani opened the pull request:
Add detail option for `_analyze` API.

This PR add these function to `_analyze` API
* output all Token's attributes
    * can specify attribute name with `attributes` parameter
    * <strike>can output short attribute name with `shor_attr_name` parameter</strike>
* output token array after each steps, that are charfilters, tokenizer, tokenfilters

Closes #11076

johtani commented:
Request example:
```json
GET localhost:9200/_analyze?pretty
{
  "text": "<text>This is troubled</text>",
  "char_filters": ["html_strip"], 
  "tokenizer": "standard", 
  "filters": ["snowball"], 
  "detail": true, 
  "short_attr_name": true, 
  "attributes": ["KeywordAttribute"]
}
```

Response example: 
```json
{
  "detail" : {
    "custom_analyzer" : true,
    "charfilters" : [ {
      "name" : "html_strip",
      "filtered_text" : [ "\nThis is troubled\n" ]
    } ],
    "tokenizer" : {
      "name" : "standard",
      "tokens" : [ {
        "token" : "This",
        "start_offset" : 6,
        "end_offset" : 10,
        "type" : "<ALPHANUM>",
        "position" : 0,
        "attributes" : { }
      }, {
        "token" : "is",
        "start_offset" : 11,
        "end_offset" : 13,
        "type" : "<ALPHANUM>",
        "position" : 1,
        "attributes" : { }
      }, {
        "token" : "troubled",
        "start_offset" : 14,
        "end_offset" : 22,
        "type" : "<ALPHANUM>",
        "position" : 2,
        "attributes" : { }
      } ]
    },
    "tokenfilters" : [ {
      "name" : "snowball",
      "tokens" : [ {
        "token" : "This",
        "start_offset" : 6,
        "end_offset" : 10,
        "type" : "<ALPHANUM>",
        "position" : 0,
        "attributes" : {
          "KeywordAttribute" : {
            "keyword" : false
          }
        }
      }, {
        "token" : "is",
        "start_offset" : 11,
        "end_offset" : 13,
        "type" : "<ALPHANUM>",
        "position" : 1,
        "attributes" : {
          "KeywordAttribute" : {
            "keyword" : false
          }
        }
      }, {
        "token" : "troubl",
        "start_offset" : 14,
        "end_offset" : 22,
        "type" : "<ALPHANUM>",
        "position" : 2,
        "attributes" : {
          "KeywordAttribute" : {
            "keyword" : false
          }
        }
      } ]
    } ]
  }
}
```

dadoonet commented on docs/reference/indices/analyze.asciidoc:
```diff
@@ -96,6 +96,24 @@ coming[2.0.0, body based parameters were added in 2.0.0]
 
+If `detail` parameter is `true`, this API output each tokens within tokenizer chain and output with all token attributes.
+It also output only token attribute specified by `attributes`.
+If `short_attr_name` is `false`, it output token attribute full name.
```
I would write this like:

If you want to get more advanced details, set `detail` to `true`  (defaults to `false`). It will output all token attributes for each token.
You can filter token attributes you want to output by setting `attributes` option.
By default, token attributes are printed using their short names but you can display the long name by setting `short_attr_name` to `false` (defaults to `true`).

dadoonet commented on docs/reference/indices/analyze.asciidoc:
```diff
@@ -96,6 +96,24 @@ coming[2.0.0, body based parameters were added in 2.0.0]
+  "attributes" : ["KeywordAttribute"],
+  "short_attr_name" : false
+}'
+--------------------------------------------------
```
I would use SENSE form

```
GET test/_analyze
{
  "field" : "obj1.field1",
  "text" : "this is a test",
  "detail" : true,
  "attributes" : ["KeywordAttribute"],
  "short_attr_name" : false
}
```

dadoonet replied:
And you can add at the end

```
--------------------------------------------------
// AUTOSENSE
```

dadoonet replied:
I think that could be nice also to have here an output example so people will know what they could expect with this new feature.

dadoonet commented on docs/reference/indices/analyze.asciidoc:
```diff
@@ -96,6 +96,24 @@ coming[2.0.0, body based parameters were added in 2.0.0]
+}'
+--------------------------------------------------
+
+coming[2.0.0, body based parameters were added in 2.0.0]
```
Is this related to the same change? Why it's not before the text you added?

johtani replied:
I think this is related the example of request. We add "body based param" style in 2.0.0.
Then, I add after the example.
If it doesn't need, I will remove this line.

dadoonet commented on rest-api-spec/api/indices.analyze.json:
```diff
@@ -44,6 +44,18 @@
         },
+        "detail": {
+          "type" : "boolean",
+          "description" : "With `true`, output each tokens within tokenizer chain with all token attributes . With `false`, (default: false)"
```
May be changing a bit the description based on the comment I wrote about the guide?
Also s/output/outputs, s/tokens/token

dadoonet commented on rest-api-spec/api/indices.analyze.json:
```diff
@@ -44,6 +44,18 @@
+        },
+        "attributes": {
+          "type" : "list",
+          "description" : "A comma-separated list of token attributes to output, this parameter works only with `detail=true`, "
```
Remove the ", " at the end

dadoonet commented on rest-api-spec/api/indices.analyze.json:
```diff
@@ -44,6 +44,18 @@
+        },
+        "short_attr_name": {
+          "type" : "boolean",
+          "description" : "With `true`, output only class name of token attribute, with `false`, output full class name of token attribute. This parameter works only with `detail=true`"
```
s/output/outputs

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -209,7 +220,25 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+
+    private List<AnalyzeResponse.AnalyzeToken> simpleAnalyze(AnalyzeRequest request, Analyzer analyzer, String field) {
+        List<AnalyzeResponse.AnalyzeToken> tokens;
+        tokens = Lists.newArrayList();
```
One single line is enough.

```java
List<AnalyzeResponse.AnalyzeToken> tokens = Lists.newArrayList();
```

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+    private DetailAnalyzeResponse detailAnalyze(AnalyzeRequest request, Analyzer analyzer, String field) {
+        DetailAnalyzeResponse detailResponse;
+        final Set<String> includeAttributes = Sets.newHashSet();
+        if (request.attributes() != null && request.attributes().length > 0) {
```
No need to check if length is > 0 because the for loop will do it. So:

```java
if (request.attributes() != null) {
```

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+
+        CustomAnalyzer customAnalyzer = null;
+        if (analyzer instanceof CustomAnalyzer) {
+            customAnalyzer = (CustomAnalyzer)analyzer;
```
Just add a space before analyzer:

```java
customAnalyzer = (CustomAnalyzer) analyzer;
```

dadoonet replied:
Small formatting issue.

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+
+        if (customAnalyzer != null) {
+            // customAnalyzer = divide charfilter, tokenizer tokenfilters
+            CharFilterFactory[] charfilterFactories = customAnalyzer.charFilters();
```
I would change for all variables here `charfilterXxxx` by `charFilterXxxx` (just like the classname)

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            // customAnalyzer = divide charfilter, tokenizer tokenfilters
+            CharFilterFactory[] charfilterFactories = customAnalyzer.charFilters();
+            TokenizerFactory tokenizerFactory = customAnalyzer.tokenizerFactory();
+            TokenFilterFactory[] tokenfilterFactories = customAnalyzer.tokenFilters();
```
same for `tokenfilter` vs `tokenFilter`

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                String charFilteredSource = text;
+
+                Reader reader = new FastStringReader(text);
+                if (charfilterFactories != null) {
```
Here you test that charfilterFactories is null. But you wrote before `Maps.newHashMapWithExpectedSize(charfilterFactories.length);`

So I suppose that charfilterFactories can not be really null?

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                    }
+                }
+
+                //analyzing only tokenizer
```
Space?
// analyzing only tokenizer

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                }
+                tokenizerTokenListCreator.analyze(tokenizer, customAnalyzer, field ,includeAttributes, request.shortAttributeName());
+
+                //analyzing each tokenfilter
```
Space?

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+
+                Reader reader = new FastStringReader(text);
+                if (charfilterFactories != null) {
+                    for (CharFilterFactory charfilter : charfilterFactories) {
```
charFilter

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                            texts = Lists.newArrayList();
+                        }
+                        texts.add(charFilteredSource);
+                        charFiltersTexts.put(charfilter.name(), texts);
```
Wondering if you want to put again the `texts` object in the map.
I would probably do something like:

```java
List<String> texts = charFiltersTexts.get(charfilter.name());
if (texts == null) {
    texts = Lists.newArrayList();
    charFiltersTexts.put(charfilter.name(), texts);
}
texts.add(charFilteredSource);
```

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                tokenizerTokenListCreator.analyze(tokenizer, customAnalyzer, field ,includeAttributes, request.shortAttributeName());
+
+                //analyzing each tokenfilter
+                if (tokenfilterFactories != null) {
```
Can not be null? Because of `Maps.newHashMapWithExpectedSize(tokenfilterFactories.length);`?

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                }
+            }
+
+            DetailAnalyzeResponse.CharFilteredText[] charfilteredLists = new DetailAnalyzeResponse.CharFilteredText[charFiltersTexts.size()];
```
charFilteredLists

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                i++;
+            }
+            DetailAnalyzeResponse.AnalyzeTokenList[] tokenfilterLists = new DetailAnalyzeResponse.AnalyzeTokenList[tokenFiltersTokenListCreator.size()];
+            i = 0;
```
I don't understand this part of the code. It sounds like to me that i is not incremented here. Is it on purpose?

johtani replied:
Good point! I found the problem about the order of tokenfilter in response.
I will fix this problem and I will change implementation.

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            }
+
+            TokenListCreator tokenListCreator = new TokenListCreator();
+            for (String text : request.text()){
```
Space missing:

```java
for (String text : request.text()) {
```

dadoonet replied:
Formatting

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            try {
+                len = input.read(buf, 0, BUFFER_SIZE);
+            } catch (IOException e) {
+                throw new ElasticsearchException("failed to analyze (charfiltering)", e);
```
charFiltering

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+        }
+
+        private AnalyzeResponse.AnalyzeToken[] getArrayTokens(){
+            return tokens.toArray(new AnalyzeResponse.AnalyzeToken[0]);
```
Might be better performance wise to use:

```java
return tokens.toArray(new AnalyzeResponse.AnalyzeToken[tokens.size()]);
```

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            for (String charFilterName : charFiltersTexts.keySet()) {
+                charfilteredLists[i] = new DetailAnalyzeResponse.CharFilteredText(
+                    charFilterName, charFiltersTexts.get(charFilterName).toArray(Strings.EMPTY_ARRAY));
+                i++;
```
This should be more efficient performance wise:

```java
List<String> strings = charFiltersTexts.get(charFilterName);
charfilteredLists[i] = new DetailAnalyzeResponse.CharFilteredText(charFilterName, strings.toArray(new String[strings.size()]));
```

dadoonet commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/analyze/RestAnalyzeAction.java:
```diff
@@ -129,7 +132,20 @@ public static void buildFromContent(BytesReference content, AnalyzeRequest analy
                             charFilters.add(parser.text());
                         }
-                        analyzeRequest.tokenFilters(charFilters.toArray(Strings.EMPTY_ARRAY));
+                        analyzeRequest.charFilters(charFilters.toArray(Strings.EMPTY_ARRAY));
```
I know that it comes from copy and paste of previous code but I think it's more efficient to do:

```java
analyzeRequest.charFilters(charFilters.toArray(new String[charFilters.size()]));
```

dadoonet commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/analyze/RestAnalyzeAction.java:
```diff
@@ -129,7 +132,20 @@ public static void buildFromContent(BytesReference content, AnalyzeRequest analy
+                            }
+                            attributes.add(parser.text());
+                        }
+                        analyzeRequest.attributes(attributes.toArray(Strings.EMPTY_ARRAY));
```
Same here:

```java
analyzeRequest.attributes(attributes.toArray(new String[attributes.size()]));
```

dadoonet commented:
@johtani I left comments. Hope this helps!

johtani commented:
@dadoonet Thanks for your comment. This was helpful for me especially about English:)
I've updated the PR with your comments.
And I found and fixed the bug about the order of tokenfilters/charfilters in the response.
Could you take another look?

johtani commented:
Add the example of response to document

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +270,230 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            TokenizerFactory tokenizerFactory = customAnalyzer.tokenizerFactory();
+            TokenFilterFactory[] tokenFilterFactories = customAnalyzer.tokenFilters();
+
+            String[][] charFiltersTexts = new String[charFilterFactories != null ? charFilterFactories.length : 0][request.text().length];
```
I did not see that much in elasticsearch code 2 dimensional arrays. Not expert in that field and what would be the consequence if any on the GC. Just wondering if it's the best way to do it. I'd like someone else to review this as well. May be @bleskes or @jpountz?

dadoonet replied:
And I wonder if it make sense to create such an array when `charFilterFactories` is null?

johtani replied:
@bleskes @jpountz Please tell me your suggestion. Should I not use 2 dimensional arrays? If not use, I can change to use ArrayList or something.

@dadoonet I create 0 length array for easy to handle output "for loop". And I think CustomAnalyzer can be set CharFilterFactories to null via Constructor.

bleskes replied:
I'm fine with double arrays, where it makes sense, i.e. we know the length of both dimensions in advance, like here. I would however use different names than i & j as indices as it can get confusing. Like textIndex & charFilterIndex ,

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +270,230 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            DetailAnalyzeResponse.CharFilteredText[] charFilteredLists = new DetailAnalyzeResponse.CharFilteredText[charFiltersTexts.length];
+            for (int j = 0; j < charFiltersTexts.length; j++) {
+                charFilteredLists[j] = new DetailAnalyzeResponse.CharFilteredText(
+                    charFilterFactories[j].name(), charFiltersTexts[j]);
```
`charFilterFactories` can be null here.

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +270,230 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+            DetailAnalyzeResponse.AnalyzeTokenList[] tokenFilterLists = new DetailAnalyzeResponse.AnalyzeTokenList[tokenFiltersTokenListCreator.length];
+            for (int j = 0; j < tokenFiltersTokenListCreator.length; j++) {
+                tokenFilterLists[j] = new DetailAnalyzeResponse.AnalyzeTokenList(
+                    tokenFilterFactories[j].name(), tokenFiltersTokenListCreator[j].getArrayTokens());
```
`tokenFilterFactories` can be null here.

dadoonet commented:
I left some other comments.

clintongormley commented:
@johtani any progress here?

johtani commented:
@clintongormley I need some other reviews for this. https://github.com/elastic/elasticsearch/pull/11660#discussion_r33843116

bleskes commented:
I replied to the comment.
> On 13 Jul 2015, at 07:31, Jun Ohtani <notifications@github.com> wrote:
> 
> @clintongormley I need some other reviews for this. #11660 (comment)
> 
> —
> Reply to this email directly or view it on GitHub.
>

johtani commented:
Thanks comments, @dadoonet @bleskes 
Fix comments.

dadoonet commented:
I left a small last comment. Not sure if you want to fix it. LGTM

johtani commented:
@dadoonet Agreed. Push fixing

dadoonet commented:
LGTM

johtani commented:
@rjernst @mikemccand @jpountz **if you have time** and interested in reviewing this, please review this. I need a second reviewer.

rjernst commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeRequest.java:
```diff
@@ -136,6 +171,9 @@ public void readFrom(StreamInput in) throws IOException {
         tokenFilters = in.readStringArray();
         charFilters = in.readStringArray();
         field = in.readOptionalString();
+        detail = in.readBoolean();
```
These new values need to be wrapped around a version check

dadoonet replied:
Did you miss Ryan's comment. You need to check the version before writing or reading the value.

rjernst commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java:
```diff
@@ -243,11 +270,234 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
+                    }
+                    currentAttributes.put(key, value);
+                    if (shortAttrName) {
+                        extendedAttributes.put(attClass.getName().substring(attClass.getName().lastIndexOf(".")+1), currentAttributes);
```
I think the api would be cleaner if we omit the classname, and just use the attribute name/value in the attributes.

johtani commented:
@rjernst Thanks for your comments. I pushed only fixing "removing classname". I think it is easy to review what is different. If it is OK, then I will rebase and fix checking version.

dadoonet commented:
@johtani What is the status on this? Do you need a final review? 

May be you should squash and rebase on master branch? Master branch changed a lot in the meantime.

For example `rest-api-spec/test/indices.analyze/10_analyze.yaml` should probably move now to `rest-api-spec/src/main/resources/rest-api-spec/test/indices.analyze/10_analyze.yaml`.

johtani commented:
@dadoonet Ah, thank you for comment. I squashed and rebased on master branch.
I need a final review!

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeResponse.java:
```diff
@@ -87,6 +112,7 @@ public void readFrom(StreamInput in) throws IOException {
             endOffset = in.readInt();
             position = in.readVInt();
             type = in.readOptionalString();
+            attributes = (Map<String, Object>) in.readGenericValue();
```
I guess that won't work if users are mixing 2.0, 2.1 and 2.2 nodes, right?

dadoonet commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeResponse.java:
```diff
@@ -141,15 +177,21 @@ public void readFrom(StreamInput in) throws IOException {
-        for (AnalyzeToken token : tokens) {
-            token.writeTo(out);
+        if (tokens != null) {
+            out.writeVInt(tokens.size());
```
I think this won't work. You are reading an int in readFrom() but you are optionally writing an int.

dadoonet commented:
I left some comments

johtani commented:
@dadoonet Thanks for reviewing. I pushed fixing.

dadoonet commented:
LGTM

johtani commented:
Thanks @dadoonet , I will merge master and v2.2.0

colings86 commented on docs/reference/indices/analyze.asciidoc:
```diff
@@ -86,6 +86,70 @@ curl -XGET 'localhost:9200/test/_analyze' -d '
+        "end_offset" : 8,
+        "type" : "<ALPHANUM>",
+        "position" : 0,
+        "keyword" : false
```
Maybe add a note here explaining that this is in the output due to the `attributes` parameter in the request. It took me a while to work out what the attributes options is actually doing

colings86 commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeRequest.java:
```diff
@@ -124,6 +147,9 @@ public ActionRequestValidationException validate() {
         }
+        if (attributes == null) {
+            validationException = addValidationError("attributes must not be null", validationException);
+        }
```
Given that attributes has a sensible default, can we not check on the setter method that the parameter is non null rather than checking here, so that way we can fail faster?

colings86 commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/analyze/RestAnalyzeAction.java:
```diff
@@ -129,7 +130,18 @@ public static void buildFromContent(BytesReference content, AnalyzeRequest analy
                         }
-                        analyzeRequest.tokenFilters(charFilters.toArray(Strings.EMPTY_ARRAY));
+                        analyzeRequest.charFilters(charFilters.toArray(new String[charFilters.size()]));
+                    } else if ("detail".equals(currentFieldName) && token == XContentParser.Token.VALUE_BOOLEAN) {
```
Can we use the ParseFieldMatcher and ParseField objects here since we should be moving all parsers over to that anyway?

colings86 commented on docs/reference/indices/analyze.asciidoc:
```diff
@@ -86,6 +86,70 @@ curl -XGET 'localhost:9200/test/_analyze' -d '
 Will cause the analysis to happen based on the analyzer configured in the
 mapping for `obj1.field1` (and if not, the default index analyzer).
 
+If you want to get more advanced details, set `detail` to `true` (defaults to `false`). It will output all token attributes for each token.
```
Might be good to put this in it's own section? Also should we mark this experimental?

colings86 commented:
@johtani I left a few comments. I also wonder if it would be more consistent with other APIs if we called the option `explain` rather than `detail` since it explains the analysis chain?

@clintongormley what do you think?

clintongormley commented:
++ for `explain`

johtani commented:
@colings86 Thanks for your comment! I will change `explain` and fixing your comment.

johtani commented:
@colings86 Changed `detail` param to `explain` and fixed your comment.
Could you review again?

colings86 commented on core/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeRequest.java:
```diff
@@ -86,6 +91,7 @@ public String tokenizer() {
     }
 
     public AnalyzeRequest tokenFilters(String... tokenFilters) {
+        assert tokenFilters != null: "token filters must not be null";
```
IMO I think this should be an if statement that throws an `IllegalArguementException` instead of an assert since a user can easily pass in null at runtime here. That way they will get an exception when its set rather than a weird exception happening when we try to use the variable. Same goes for `charFilters` and `attributes` below

colings86 commented:
@johtani I left one minor comment but LGTM

johtani commented:
Closed by master : fab44398d9d48f12319bc018d4b436f723b6508e 
2.x : f427cf4ae459b864894c8ba1a3eff589ad5248a0
